### PR TITLE
C++ API: torch::nn::BatchNorm{2,3}d

### PR DESCRIPTION
**Pull request by nuka137**
Add torch::nn::BatchNorm{2,3}d module and functional support for the C++ API.

Related Issue: #25883 #28176

Reviewer: @yf225

### CODE REVIEW

**Review by yf225 — COMMENTED** [review]
@nuka137 Thanks so much for the awesome work and thorough tests! I have one very minor comment.

File: torch/csrc/api/src/nn/modules/batchnorm.cpp [hunk 1]
```diff
@@ -168,5 +168,21 @@ void BatchNorm1dImpl::_check_input_dim(const Tensor& input) {
...
+
+template class BatchNormImplBase<2, BatchNorm2dImpl>;
+
+void BatchNorm3dImpl::_check_input_dim(const Tensor& input) {
+  TORCH_CHECK(
+      input.dim() == 5,
+      "expected 5D input (got ", input.dim(), "D input)");
+}
+
+template class BatchNormImplBase<3, BatchNorm3dImpl>;
```
[the remark this answers is not in the document]
Comment (nuka137) [reply torch/csrc/api/src/nn/modules/batchnorm.cpp]: Thanks. Fixed it.

**Comment by nuka137** [comment]
@yf225 

Thanks for your review. I fixed the source code you commented.
Could you check it again?

**Review by yf225 — APPROVED** [review]
@nuka137 Thanks so much for the awesome work!

**Comment by ezyang** [comment]
TEsts are failing:

```

Nov 01 17:32:00 [ RUN      ] ModulesTest.BatchNorm2dStateful
Nov 01 17:32:00 /var/lib/jenkins/workspace/test/cpp/api/modules.cpp:1224: Failure
Nov 01 17:32:00 Expected equality of these values:
Nov 01 17:32:00   bn->num_batches_tracked.dim()
Nov 01 17:32:00     Which is: 0
Nov 01 17:32:00   1
```

**Comment by yf225** [comment]
@nuka137 Sorry it was my fault that causes this PR to be reverted - I will re-open this PR and make fixes directly to this PR, and then makes sure it passes CI before merging again.

**Comment by jerryzh168** [comment]
Can we split aten function to aten::batch_norm_1d, aten::batch_norm_2d and aten::batch_norm_3d and move the dimension check inside the aten function? This will simplify the other works that needs to match batchnorm module or op. cc @ZolotukhinM

**Comment by yf225** [comment]
@jerryzh168 Yes I think that would be beneficial - the C++ API implementation here is just strictly following Python API implementation, and we can change it once the Python implementation is changed.
